common/telemetry/env.go

```diff
@@ -0,0 +1,107 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package telemetry
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	otelsdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.temporal.io/server/common/primitives"
+	"golang.org/x/exp/maps"
+)
+
+var (
+	unsupportedTraceExporter         = errors.New("unsupported OTEL exporter")
+	unsupportedTraceExporterProtocol = errors.New("unsupported OTEL exporter protocol")
+)
+
+const (
+	OtelServiceNameEnvKey                = "OTEL_SERVICE_NAME"
+	OtelTracesExporterTypesEnvKey        = "OTEL_TRACES_EXPORTER"
+	OtelTracesOtlpExporterType           = SpanExporterType("otlp")
+	OtelExporterOtlpTracesProtocolEnvKey = "OTEL_EXPORTER_OTLP_TRACES_PROTOCOL"
+	OtelExporterOtlpTracesGrcpProtocol   = "grpc"
+)
+
+type envVarLookup = func(string) (string, bool)
+
+// SupplementTraceExportersFromEnv adds any OTEL span exporters configured through environment variables
+// to a given set of exporters. It does not override an existing exporter with the same type.
+func SupplementTraceExportersFromEnv(
+	exporters map[SpanExporterType]otelsdktrace.SpanExporter,
+	envVars envVarLookup,
+) error {
+	exporterTypes, ok := envVars(OtelTracesExporterTypesEnvKey)
+	if !ok {
+		return nil
+	}
+
+	supplements := map[SpanExporterType]otelsdktrace.SpanExporter{}
+	for _, exporterType := range strings.Split(exporterTypes, ",") {
+		switch SpanExporterType(exporterType) {
+		case OtelTracesOtlpExporterType:
+			if _, exists := exporters[OtelTracesOtlpExporterType]; exists {
+				continue
+			}
+
+			// only grpc is supported; fail if user requests a different protocol
+			if protocol, exists := envVars(OtelExporterOtlpTracesProtocolEnvKey); exists {
```

Comment:
question: could this if be moved out of the for loop?

Reply:
I don't think it can, because it's checking the `OTEL_EXPORTER_OTLP_TRACES_PROTOCOL` which is a sub-setting for `OTEL_TRACES_EXPORTER` (which the outer switch statement is checking for).